### ISSUE TITLE
Pkg.build failure does not propagate

### ISSUE THREAD

**Issue by JeffBezanson** [JuliaLang/julia#15045]
When `Pkg.build` fails it prints a message, but does not propagate an exception (or result in a non-0 exit code when run from a script). I know this was probably done intentionally to allow trying to build multiple packages with minimal blockage, but I don't think it's a good default. It's very surprising to have, say, a VM setup script appear to succeed but have package build failures.

**Comment by tkelman** [comment]
This sounds like a regression, on 0.4 I'm pretty sure I would see exceptions and/or nonzero exit status for this?

**Comment by JeffBezanson** [comment]
I get the same behavior in 0.4.2 at least.

**Comment by tkelman** [comment]
Hm, can we come up with a self contained test case for reproduction then? Docker container that doesn't have the right prereqs installed, or something like that?

**Comment by JeffBezanson** [comment]
Could we put it in test-pkg, where IIUC we make some small mock packages, and could give one of them a build script that throws an error?

**Comment by tkelman** [comment]
Yes that could also work. I know I've seen this but it's usually something I fix outside of julia and don't always write down the conditions that were responsible. But I'd like to see how any change impacts real world cases of this, in addition to putting a synthetic case in the test suite.

**Comment by stevengj** [comment]
I think this is a longstanding behavior.  @StefanKarpinski wrote in 51376f6370732cc84a02227d5471862816de8d69 way back in 0.2:
> Philosophically, this change takes the approach that as far as the
package installation machinery is concerned, build failures aren't
really it's problem – it will run the build script for a package
when that package is installed or updated, but if the build script
fails, the package is still considered to be installed. A big, red
warning is printed for the user so that they know the package is
not likely to work, but the package is left installed so that they
can go into its directory and try to fix it. This approach seems
likely to make installing packages with binary dependencies a lot
less annoying.

**Comment by tkelman** [comment]
Though a `Pkg.build` called directly could potentially be made to have different exception throwing behavior than a `Pkg.build` caused indirectly inside `Pkg.add`.

**Comment by StefanKarpinski** [comment]
Yes, I think what Tony suggested is spot on.

**Comment by JeffBezanson** [comment]
It seems to me it's possible to preserve the properties the package manager wants while still propagating an exception in all cases. All it needs to do is run build scripts after everything is installed, and not make any state changes depending on whether a build fails.
